src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageVulnerabilityService.cs

```diff
@@ -17,26 +18,28 @@ namespace NuGet.PackageManagement.VisualStudio
     public class PackageVulnerabilityService : IPackageVulnerabilityService
     {
         private readonly IEnumerable<SourceRepository> _sourceRepositories;
-        private GetVulnerabilityInfoResult _vulnerabilities;
+        private AsyncLazy<GetVulnerabilityInfoResult> _vulnerabilities;
         private INuGetUILogger _logger;
+        private readonly CancellationToken _cancellationToken;
 
-        public PackageVulnerabilityService(IEnumerable<SourceRepository> sourceRepositories, INuGetUILogger logger)
+        public PackageVulnerabilityService(IEnumerable<SourceRepository> sourceRepositories, INuGetUILogger logger, CancellationToken cancellationToken)
         {
             _sourceRepositories = sourceRepositories ?? throw new ArgumentNullException(nameof(sourceRepositories));
-            _logger = logger ?? throw new ArgumentNullException(nameof(logger)); ;
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _cancellationToken = cancellationToken;
+            _vulnerabilities = new AsyncLazy<GetVulnerabilityInfoResult>(async () => await GetAllVulnerabilityDataAsync(cancellationToken));
         }
 
-        public async Task<List<PackageVulnerabilityMetadataContextInfo>> GetVulnerabilityInfoAsync(PackageIdentity packageId, CancellationToken cancellationToken)
+        public async Task<List<PackageVulnerabilityMetadataContextInfo>> GetVulnerabilityInfoAsync(PackageIdentity packageId)
         {
-            _vulnerabilities = await GetAllVulnerabilityDataAsync(cancellationToken);
             IEnumerable<PackageVulnerabilityInfo> packageVulnerabilities = Enumerable.Empty<PackageVulnerabilityInfo>();
-
-            if (_vulnerabilities?.Exceptions is not null)
+            GetVulnerabilityInfoResult vulnerabilityInfoResult = await _vulnerabilities;
+            if (vulnerabilityInfoResult?.Exceptions is not null)
             {
-                ReplayErrors(_vulnerabilities.Exceptions);
+                ReplayErrors(vulnerabilityInfoResult.Exceptions);
             }
 
-            IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> allVulnerabilities = _vulnerabilities?.KnownVulnerabilities;
+            IReadOnlyList<IReadOnlyDictionary<string, IReadOnlyList<PackageVulnerabilityInfo>>> allVulnerabilities = vulnerabilityInfoResult?.KnownVulnerabilities;
```

Comment:
Starting new thread for this @jebriede comment: 

> The problem we are fixing stems from how we are calling into GetVulnerabilityInfoAsync on every call which in turn calls into GetVulnerabilityInfoAsync for each source repo

With the addition of AuditSources feature, I think we shouldn't be calling `GetVulnerabilityInfoAsync ` for each source repo since right now only nuget.org implements vulnerable info. 

Also, I think we have a bug here on how we merge this vulnerability info in the PM UI, if the user has selected 'All' sources we are not showing vulnerability info if the user has nuget.org + another source.

**I'm not saying we should fix it in this PR** but this PR is related to this area, and we could identify how to fix it

Reply:
> Also, I think we have a bug here on how we merge this vulnerability info in the PM UI, if the user has selected 'All' sources we are not showing vulnerability info if the user has nuget.org + another source.

Bug or missing part to the feature? That behavior was not well defined when we implemented this. With the addition of audit sources, we can make a better decision of how to handle this scenario.

> I'm not saying we should fix it in this PR but this PR is related to this area, and we could identify how to fix it

Agreed, this is out of scope. Please file an issue if one does not already exist.

Reply:
The vulnerability info provider accounts for that.
It's a completely new resource, so if a source doesn't support it'll return nothing. 

The thing is in order to figure out if a source supports it, you need to look at the index.json and that's what this code currently does.

Reply:
Thanks for the context. I'll close this conversation since it's not relevant to the actual PR, there is a bug for this https://github.com/NuGet/Home/issues/13954